mempool/src/shared_mempool.rs

```diff
@@ -277,65 +405,70 @@ async fn inbound_network_task<V>(
-        trace!("SharedMempoolEvent::NetworkEvent::{:?}", event);
-        match event {
-            Ok(network_event) => match network_event {
-                Event::NewPeer(peer_id) => {
-                    OP_COUNTERS.inc("smp.event.new_peer");
-                    new_peer(&peer_info, peer_id);
-                    notify_subscribers(SharedMempoolNotification::PeerStateChange, &subscribers);
-                }
-                Event::LostPeer(peer_id) => {
-                    OP_COUNTERS.inc("smp.event.lost_peer");
-                    lost_peer(&peer_info, peer_id);
-                    notify_subscribers(SharedMempoolNotification::PeerStateChange, &subscribers);
-                }
-                Event::Message((peer_id, msg)) => {
-                    OP_COUNTERS.inc("smp.event.message");
-                    let transactions: Vec<_> = msg
-                        .transactions
-                        .clone()
-                        .into_iter()
-                        .filter_map(|txn| match SignedTransaction::try_from(txn) {
-                            Ok(t) => Some(t),
-                            Err(e) => {
-                                security_log(SecurityEvent::InvalidTransactionMP)
-                                    .error(&e)
-                                    .data(&msg)
-                                    .log();
-                                None
-                            }
-                        })
-                        .collect();
-                    OP_COUNTERS.inc_by(
-                        &format!("smp.transactions.received.{:?}", peer_id),
-                        transactions.len(),
-                    );
-                    bounded_executor
-                        .spawn(process_incoming_transactions(
-                            smp.clone(),
-                            peer_id,
-                            transactions,
-                        ))
-                        .await;
-                }
-                _ => {
-                    security_log(SecurityEvent::InvalidNetworkEventMP)
-                        .error("UnexpectedNetworkEvent")
-                        .data(&network_event)
-                        .log();
-                    debug_assert!(false, "Unexpected network event");
+    let mut events = select_all(network_events).fuse();
+
+    loop {
+        ::futures::select! {
+            // client events
+            (mut msg, callback) = client_events.select_next_some() => {
+                bounded_executor
+                .spawn(process_client_transaction_submission(
+                    smp.clone(),
+                    msg,
+                    callback,
+                ))
+                .await;
+            }
+            event = events.select_next_some() => {
+                match event {
+                    Ok(network_event) => match network_event {
+                        Event::NewPeer(peer_id) => {
```

Comment:
maybe I misread it, but it looks like messages will get broadcasted in all directions through all networks that possible. That's not something we want to. Can we change logic to only do it for upstream?